lib/FusionInventory/Agent/Task/Inventory/Linux/Storages.pm

```diff
@@ -73,11 +97,10 @@ sub _getDevices {
         # the hash keys are function references from %sources
         my %info;
 
-        for my $field (keys %sources) {
-            next if defined $device->{$field}
-                && !($field eq 'MANUFACTURER' && $device->{$field} eq 'ATA');
+        while (my ($field, $data) = each %sources) {
+            next if defined $device->{$field} && $data->{next_if}->($device->{$field});
 
-            for my $sub (@{$sources{$field}}) {
+            for my $sub (@{$data->{subs}}) {
```

Comment:
To follow my previous comment, here is what I'm thinking about without the need of %sources. I didn't try this code so maybe I'm wrong, but can you try this way of doing ?
```perl
        # By default, we will get other info from smartctl and then from hdparm
        my $default_subs = [ \&getInfoFromSmartctl, \&_getHdparmInfo ];
        foreach my $field (qw(DESCRIPTION DISKSIZE FIRMWARE INTERFACE MANUFACTURER MODEL WWN)) {
            my $subs = $default_subs;
            if ($field eq 'MANUFACTURER') {
                # Try to update manufacturer if set to SATA
                $subs = [ \&getInfoFromSmartctl ];
                next if defined($device->{$field}) && $device->{$field} ne 'SATA';
            } else {
                # eventually add missing fields but overwrite MODEL with whatever returned from subs
                next if defined $device->{$field} && $field ne 'MODEL';
            }
            
            foreach my $sub (@{$subs}) {
```

Reply:
Done.

Btw, "SATA" check makes no sense here. "ATA" check is for cases like:
```
$ cat /sys/block/sda/device/vendor
ATA
```